### PR TITLE
Feature/welcomewidget

### PR DESCRIPTION
**Pull request by martinfalk**
* Added a search for finding workspaces in the welcome widget.
* replaced the clear icon in search fields (e.g. processor list)

**Searching for boron**
image

**Before**
image

**After**
image

### CODE REVIEW

File: include/inviwo/qt/editor/filetreewidget.h [hunk 1]
```diff
@@ -34,55 +34,133 @@
...
     QIcon fileIcon_;
 };
 
+class IVW_QTEDITOR_API TreeItem {
+public:
+    explicit TreeItem(TreeItem* parent = nullptr);
+    TreeItem(const QString& caption, FileTreeWidget::ListElemType type, TreeItem* parent = nullptr);
+    TreeItem(const QIcon& icon, const std::string& filename, bool isExample = false,
+             TreeItem* parent = nullptr);
+    virtual ~TreeItem();
```
Comment (petersteneteg) [inline include/inviwo/qt/editor/filetreewidget.h]: why virtual?

File: include/inviwo/qt/editor/filetreewidget.h [hunk 2]
```diff
@@ -34,55 +34,133 @@
...
 
-class IVW_QTEDITOR_API FileTreeWidget : public QTreeWidget {
+class IVW_QTEDITOR_API FileTreeWidget : public QTreeView {
 #include <warn/push>
 #include <warn/ignore/all>
     Q_OBJECT
 #include <warn/pop>
 public:
-    enum ListElemType { File = 1, Section, SubSection };
+    enum ListElemType { File = 1, Section, SubSection, None };
```
Comment (petersteneteg) [inline include/inviwo/qt/editor/filetreewidget.h]: enum class

File: include/inviwo/qt/editor/filetreewidget.h [hunk 3]
```diff
@@ -34,55 +34,133 @@
...
+    virtual FileTreeWidget::ListElemType type() const;
+
+    void setData(const QString& caption, FileTreeWidget::ListElemType type);
+    void setData(const QIcon& icon, const std::string& filename, bool isExample);
+
+private:
+    TreeItem* parent_;
+    std::vector<TreeItem*> childItems_;
+
+    FileTreeWidget::ListElemType type_ = FileTreeWidget::ListElemType::None;
```
Comment (petersteneteg) [inline include/inviwo/qt/editor/filetreewidget.h]: Could use variant here...

File: include/inviwo/qt/editor/filetreewidget.h [hunk 4]
```diff
@@ -34,55 +34,133 @@
...
+    TreeItem* parent() const;
+
+    virtual QVariant data(int column, int role) const;
+    virtual FileTreeWidget::ListElemType type() const;
+
+    void setData(const QString& caption, FileTreeWidget::ListElemType type);
+    void setData(const QIcon& icon, const std::string& filename, bool isExample);
+
+private:
+    TreeItem* parent_;
```
Comment (petersteneteg) [inline include/inviwo/qt/editor/filetreewidget.h]: ownership?

File: src/qt/editor/filetreewidget.cpp [hunk 5]
```diff
@@ -43,65 +44,360 @@
...
+
+    for (int row = 0; row < count; ++row) delete childItems_[position];
+    childItems_.erase(childItems_.begin() + position);
+
+    return true;
+}
+
+void TreeItem::removeChildren() {
+    for (auto c : childItems_) {
+        delete c;
```
Comment (petersteneteg) [inline src/qt/editor/filetreewidget.cpp]: why not using unique_ptr?

File: src/qt/editor/filetreewidget.cpp [hunk 6]
```diff
@@ -43,65 +44,360 @@
...
+    file_ = utilqt::toQString(filesystem::getFileNameWithExtension(filename));
+    path_ = utilqt::toQString(filesystem::getFileDirectory(filename));
+    isExample_ = isExample;
+
+    if (path_.isEmpty()) {
+        path_ = ".";
+    }
+}
+
+TreeModel::TreeModel(QObject* parent) : QAbstractItemModel(parent), root_{new TreeItem(nullptr)} {}
```
Comment (petersteneteg) [inline src/qt/editor/filetreewidget.cpp]: use unique_ptr?

File: src/qt/editor/filetreewidget.cpp [hunk 7]
```diff
@@ -43,65 +44,360 @@
...
+    endRemoveRows();
+
+    return success;
+}
+
+void TreeModel::updateCategory(TreeItem* item, std::vector<TreeItem*> children) {
+    QModelIndex index = getIndex(item);
+
+    beginRemoveRows(index, 0, item->childCount());
+
```
Comment (petersteneteg) [inline src/qt/editor/filetreewidget.cpp]: nothing? remove/comment?

File: src/qt/editor/filetreewidget.cpp [hunk 8]
```diff
@@ -43,65 +44,360 @@
 #include <QPainter>
 #include <QStyle>
 #include <QApplication>
+#include <QSortFilterProxyModel>
 
 #include <warn/pop>
 
 namespace inviwo {
 
+TreeItem::TreeItem(TreeItem* parent) : parent_{parent} {}
```
Comment (petersteneteg) [inline src/qt/editor/filetreewidget.cpp]: Might put the model and tree in separate files since they are quite large...

**Review by petersteneteg — APPROVED** [review]
Sweet, have some minor potential improvements, depending on what qt allows...

File: include/inviwo/qt/editor/filetreewidget.h (hunk 1 above)
Comment (martinfalk) [reply include/inviwo/qt/editor/filetreewidget.h]: left-over from some experimenting

**Review by petersteneteg — APPROVED** [review]

**Pull request merged**